RFCs/FS-1060-nullable-reference-types.md

```diff
@@ -0,0 +1,1126 @@
+>
+> `[NonNullTypes(true|false)]` is not synthesized by the compiler. If the attribute is used explicitly in source, the type declaration must be provided explicitly to the compilation.
+
+In other words, this attribute specifies a way to mark a class, constructor, delegate, enum, event, field, interface, method, module, property, or struct as having nullability expressable or not in a containing scope.
+
+This attribute could potentially be used in F# to allow for opt-in/opt-out nullability at a fine-grained level.
+
+F# will respect this attribute, with the `true` caseindicating that the scope distinguishes between nullable and non-nullable reference types.
+
+F# will treat scopes with the `false` case as if its contained reference types are all nullable reference types.
+
+In other words, if we cannot be sure that a given reference type is non-nullable, we assume it is nullable.
+
+As a final note, this attribute _can_ be abused. For example, if a method is annotated with `[NonNullTypes(true)]` and does not actually perform a `null` check, further code can still produce a `NullReferenceException` and not have a warning associated with it.
+
+#### Null obliviousness
+
+Nullability obliviousness is a concept that C# 8.0 has. A null-oblivious type is one that no assumptions can be made about. Once assigned to a nullable or non-nullable variable, it is treated as if it is nullable or non-nullable, respectively.
+
+**F# may not have this concept. Reference types we cannot determine to be non-nullable will be assumed to be nullable.**
+
+This is controversial, because we are making an assumption about code that we do not really know about:
+
+* The code could not be handling `null` at all, and thus could either produce a `NullReferenceException` or produce a value that is nullable
+* The code could be handling `null` just fine, and could pass back a reference type that can never be `null`
+* We have no way to tell if any of the previous two points are true
+
+In practice, this could mean emitting thousands of warnings for a single project. Although emitting warnings for unsafe null usage is a feature, if the number of warnings gets too high, people will find it invasive.
+
+To remain "true" to the original code, we could conceivably introduce null obliviousness and thus treat the value coming out of a component as either nullable or non-nullable, depending on how we assign it.
+
+However, this is not actually possible in F#, because the majority of F# code uses type inference to infer a type. Null obliviousness would require explicit type annotations to work as we intend it! Failing the propagation of a null-oblivous type through type inference (see [Type Inferece](FS-1060-nullabl-reference-types.md#type-inference)), we must assume nullability to remain safe. (**NOTE:** this is also under discussion for C# with `var x = ...`).
```

Comment:
> Failing the propagation of a null-oblivous type through type inference (see [Type Inferece](FS-1060-nullabl-reference-types.md#type-inference)), we must assume nullability to remain safe. (**NOTE:** this is also under discussion for C# with `var x = ...`).

Per the above, I don't really understand this comment. Perhaps it will become clear later